File: 1/2021.07.15/253.Meeting Rooms II/java/solution.java
public class Solution {
    public int minMeetingRooms(int[][] intervals) {
        Arrays.sort(intervals, Comparator.comparingInt(a -> a[0]));
        PriorityQueue<Integer> heap = new PriorityQueue<>();
        for (int[] interval : intervals) {
            if (!heap.isEmpty() && interval[0] >= heap.peek()) {
                heap.poll();
            }
            heap.offer(interval[1]);
        }
        return heap.size();
    }
}

File: 1/2021.07.15/253.Meeting Rooms II/python/solution.py
from heapq import heappush, heappop
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        intervals.sort(key=lambda x : x[0])
        heap = []
        for interval in intervals:
            if heap and interval[0] >= heap[0]:
                heappop(heap)
            heappush(heap, interval[1])
        return len(heap)
